File: go/conformance/adapter.go
package conformance

import "context"

// The contract between this runner and an SDK under test. Kept this thin on
// purpose: anything richer would let a case assert something idiomatic, and
// cases must only ever verify observable behaviour.
type ReportedError struct {
	// Code is the envelope error_code when the failure came from the API.
	Code    int
	HasCode bool
	Message string
}

type BotUnderTest interface {
	// Start begins polling and returns immediately.
	Start(ctx context.Context)
	// StartAgain calls Start a second time and reports what happened. It must
	// not block: a bot that is already running has to reject immediately, which
	// is the whole of G1.
	StartAgain() error
	// Stop cancels any in-flight poll and returns once the bot is idle.
	Stop()
	// StoppedItself reports a bot that ended on a terminal error rather than on
	// Stop.
	StoppedItself() bool
	HandlerRuns() map[int64]int
	Errors() []ReportedError
	Warnings() []string
}

type BotOptions struct {
	BaseURL        string
	Token          string
	Limit          int
	TimeoutSeconds int
	Handler        HandlerSpec
}

type Factory func(BotOptions) BotUnderTest

File: js/conformance/adapter.ts
import type { HandlerSpec } from "./types.js";

/**
 * The contract between the conformance runner and an SDK under test.
 *
 * The runner is written against this interface and nothing else, so it exists
 * before the SDK does and tells the SDK what to expose. Keeping it this thin is
 * deliberate: anything richer would let a case assert something idiomatic, and
 * cases must only ever verify observable behaviour (§3 of 01-organization.md).
 */
export interface ReportedError {
  /** Envelope error_code when the failure came from the API, else undefined. */
  code?: number;
  message: string;
}

export interface BotUnderTest {
  /** Resolves once polling has begun. Does not wait for the bot to finish. */
  start(): Promise<void>;
  /** Aborts any in-flight poll and returns once the bot is idle. */
  stop(): Promise<void>;

  /** True when the bot stopped on its own — a terminal error, not stop(). */
  readonly stoppedItself: boolean;
  /** How many times the handler ran, keyed by update_id. */
  readonly handlerRuns: ReadonlyMap<number, number>;
  /** Errors surfaced to the author, in order, already redacted. */
  readonly errors: readonly ReportedError[];
  /** Warnings surfaced to the author, in order. */
  readonly warnings: readonly string[];
}

export interface BotUnderTestOptions {
  baseUrl: string;
  token: string;
  limit?: number;
  timeoutSeconds?: number;
  handler: HandlerSpec;
}

export interface BotFactory {
  create(options: BotUnderTestOptions): BotUnderTest;
}

/**
 * Where the runner looks for the SDK's adapter. The SDK provides this module;
 * until it exists, `--factory` points the runner at a fixture instead.
 */
export const DEFAULT_FACTORY = "../src/conformance-adapter.js";
